src/analyzer/Analyzer.cc

```diff
@@ -909,3 +910,41 @@ void SupportAnalyzer::ForwardUndelivered(uint64_t seq, int len, bool is_orig)
 	}
 
 	} // namespace zeek::analyzer
+
+TEST_SUITE("Analyzer management")
+	{
+	TEST_CASE("Re-add analyzer after removal")
+		{
+		// This test tries to reactivate an analyzer which was previously removed.
+		// It's a regression test for #2801.
+		REQUIRE(zeek::analyzer_mgr);
+
+		zeek::Packet p;
+		zeek::ConnTuple t;
+		auto conn = std::make_unique<zeek::Connection>(zeek::detail::ConnKey(t), 0, &t, 0, &p);
+		auto* tcp = new zeek::packet_analysis::TCP::TCPSessionAdapter(conn.get());
+		conn->SetSessionAdapter(tcp, nullptr);
+
+		auto a = zeek::analyzer_mgr->InstantiateAnalyzer("SSH", conn.get());
+		REQUIRE(a);
+		auto b1 = zeek::analyzer_mgr->InstantiateAnalyzer("IMAP", a->Conn());
+		REQUIRE(b1);
+
+		tcp->AddChildAnalyzer(a);
+		a->AddChildAnalyzer(b1);
+
+		CHECK(conn->FindAnalyzer("SSH"));
+		CHECK(conn->FindAnalyzer("IMAP"));
+
+		a->RemoveChildAnalyzer(b1);
+
+		CHECK(! conn->FindAnalyzer("IMAP"));
+
+		auto b2 = zeek::analyzer_mgr->InstantiateAnalyzer("IMAP", a->Conn());
+		REQUIRE(b2);
+
+		REQUIRE(a->AddChildAnalyzer(b2));
+		CHECK(conn->FindAnalyzer("IMAP"));
+		conn->Done();
```

Comment:
Ah, this is the magic bit I was missing with the sketch(y) unit test in the original report.

Maybe more as a comment, having to call a `Done` method or else there would be a random segfault isn't ideal in a unit test since e.g., any `REQUIRE` failure above would skip the rest of the test, which would then lead to weird issues elsewhere. A less fragile approach would be to use RAII to clean up the connection automatically, e.g.,

```cpp
		auto conn = std::unique_ptr<zeek::Connection, void (*)(zeek::Connection*)>(
			new zeek::Connection(zeek::detail::ConnKey(t), 0, &t, 0, &p),
			[](auto* c)
			{
				c->Done();
				delete c;
			});
```

Reply:
Yeah, was a bit of trial and error until the unit test stopped crashing. Agree in general on the comment, but not sure it's worth it: I'm sure there are more cleanup issues in all of this artificial setting up. it's a broader issue: Zeek's classes & state isn't built for operating outside the full system. And at least it'll cause problem on failures only, so not too worried here.